### PR TITLE
Praj/sustain 633

### PR DESCRIPTION
**Pull request by PrajaktaPurohit**
##Build:
http://wilson.cd.chef.co/job/chef-server-trigger-ad_hoc/6/downstreambuildview/

This updates ruby to 2.4.1, rest_client to 2.0.2 and uses a monkeypatch for the connect() method in http/net.
The bug with IPV6 and Net::HTTP is seen when chef-pedant on IPV6 (any app over ssl on IPV6) connects to Net::HTTP via a proxy.
The reason we see this bug is because https://github.com/ruby/ruby/blob/ruby_2_4/lib/net/http.rb#L918 does not include the '[' ']' literals.
And cannot be used for forwarding over proxy. This is not fixed upstream and hence we need the monkeypatch.

Users using opsworks(probably) and facebook would be broken without this monkeypatch.

##Testing for the monkeypatch:
Terminal 1
```
root@api:/host/oc-chef-pedant/lib/pedant/core_ext# nc -6 -l -k :: 8080
GET / HTTP/1.1
Accept-Encoding: gzip;q=1.0,deflate;q=0.6,identity;q=0.3
Accept: */*
User-Agent: Ruby
Host: ::1:8080

??[R7??!??qW?xpe?&5Ć
                    "9{n??q?>?+?/?,?0?????#?'?	??$?(?
?gk39@j28??<=/5?::1
```
Terminal2:
```
root@api:/host/oc-chef-pedant# bundle exec irb
irb(main):001:0> require_relative 'lib/pedant/core_ext/net_http.rb'
=> true

 #some output truncated

irb(main):004:0> Net::HTTP.get(URI("http://[::1]:8080"))
Errno::ECONNREFUSED: Failed to open TCP connection to ::1:8080 (Connection refused - connect(2) for "::1" port 8080)
	from /host/oc-chef-pedant/lib/pedant/core_ext/net_http.rb:21:in `rescue in block in connect'
	from /host/oc-chef-pedant/lib/pedant/core_ext/net_http.rb:18:in `block in connect'
	from /opt/opscode/embedded/lib/ruby/2.4.0/timeout.rb:93:in `block in timeout'
	from /opt/opscode/embedded/lib/ruby/2.4.0/timeout.rb:103:in `timeout'
	from /host/oc-chef-pedant/lib/pedant/core_ext/net_http.rb:17:in `connect'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:1485:in `begin_transport'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:1442:in `transport_request'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:1416:in `request'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:1317:in `request_get'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:487:in `block in get_response'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:877:in `start'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:608:in `start'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:485:in `get_response'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:462:in `get'
	from (irb):4
	from /opt/opscode/embedded/bin/irb:11:in `<top (required)>'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/cli/exec.rb:75:in `load'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/cli/exec.rb:75:in `kernel_load'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/cli/exec.rb:28:in `run'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/cli.rb:424:in `exec'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/cli.rb:27:in `dispatch'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/cli.rb:18:in `start'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/exe/bundle:30:in `block in <top (required)>'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/friendly_errors.rb:122:in `with_friendly_errors'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/exe/bundle:22:in `<top (required)>'
	from /opt/opscode/embedded/bin/bundle:23:in `load'
	from /opt/opscode/embedded/bin/bundle:23:in `<main>'
irb(main):005:0> Net::HTTP.get(URI("https://[::1]:8080"))
Errno::ECONNREFUSED: Failed to open TCP connection to ::1:8080 (Connection refused - connect(2) for "::1" port 8080)
	from /host/oc-chef-pedant/lib/pedant/core_ext/net_http.rb:21:in `rescue in block in connect'
	from /host/oc-chef-pedant/lib/pedant/core_ext/net_http.rb:18:in `block in connect'
	from /opt/opscode/embedded/lib/ruby/2.4.0/timeout.rb:93:in `block in timeout'
	from /opt/opscode/embedded/lib/ruby/2.4.0/timeout.rb:103:in `timeout'
	from /host/oc-chef-pedant/lib/pedant/core_ext/net_http.rb:17:in `connect'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:887:in `do_start'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:876:in `start'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:608:in `start'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:485:in `get_response'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:462:in `get'
	from (irb):5
	from /opt/opscode/embedded/bin/irb:11:in `<top (required)>'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/cli/exec.rb:75:in `load'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/cli/exec.rb:75:in `kernel_load'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/cli/exec.rb:28:in `run'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/cli.rb:424:in `exec'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/cli.rb:27:in `dispatch'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/cli.rb:18:in `start'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/exe/bundle:30:in `block in <top (required)>'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/lib/bundler/friendly_errors.rb:122:in `with_friendly_errors'
	from /opt/opscode/embedded/lib/ruby/gems/2.4.0/gems/bundler-1.16.0/exe/bundle:22:in `<top (required)>'
	from /opt/opscode/embedded/bin/bundle:23:in `load'
	from /opt/opscode/embedded/bin/bundle:23:in `<main>'
irb(main):006:0> 
irb(main):007:0* Net::HTTP.get(URI("https://[::1]:8080"))
ssl true
proxy false
Net::OpenTimeout: Net::OpenTimeout
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/protocol.rb:41:in `ssl_socket_connect'
	from /host/oc-chef-pedant/lib/pedant/core_ext/net_http.rb:66:in `connect'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:887:in `do_start'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:876:in `start'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:608:in `start'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:485:in `get_response'
	from /opt/opscode/embedded/lib/ruby/2.4.0/net/http.rb:462:in `get'
	from (irb):7

```

From the stack  trace we can see that the code in the `/host/oc-chef-pedant/lib/pedant/core_ext/net_http.rb:66:in `connect'` that has the monkeypatch is called. 

## Concerns:
The headers in netcat output in terminal1 should not have shown the Host: ::1:8080.
It should have not had any headers for the Host since there is no proxy and first case is non-ssl

## TODO:
[x] Travis failure due to jazz_hands gem

There was another bug for IPV6 with Net::Http because the TCP request could not be made. That is fixed with ruby to 2.4.1, rest_client to 2.0.2.

### CODE REVIEW

**Comment by PrajaktaPurohit** [comment]
Ref: https://github.com/chef/chef-server/pull/1323

**Comment by srenatus** [comment]
@PrajaktaPurohit Thanks for working on this. Just while you have the context around this, could you elaborate on this, maybe in one of the commit messages?

> There was another bug for IPV6 with Net::Http because the TCP request could not be made. That is fixed with ruby to 2.4.1, rest_client to 2.0.2.

On the other hand, I don't see us realistically have a reason to go back to an earlier version of `rest-client`, so I don't think this is urgent.

File: oc-chef-pedant/lib/pedant/core_ext/net_http.rb [hunk 1]
```diff
@@ -13,85 +13,108 @@
...
-    # domain names, it should fail on a name lookup.
-    # By the time this is called, the socket should be
-    # open, so whatever address is being used would
-    # technically be a valid one.
-    def ipv6_address?(addr)
-      return true if addr =~ /:/
-      return false
+    def host_header		
+      if ipv6_address?(@address)		
+        "Host: [#{@address}]:#{@port}"
```
Comment (markan) [inline oc-chef-pedant/lib/pedant/core_ext/net_http.rb]: In line 49 above we also print Host: so this probably needs to be fixed so we don't print Host: Host: addr:port

Comment (PrajaktaPurohit) [reply oc-chef-pedant/lib/pedant/core_ext/net_http.rb]: done!

**Comment by PrajaktaPurohit** [comment]
@srenatus - for sure! updated the details in a commit message!

**Pull request merged**